pkg/gktest/runner_test.go

```diff
@@ -641,3 +609,344 @@ func TestRunner_Run_ClientError(t *testing.T) {
 		t.Error(diff)
 	}
 }
+
+func TestRunner_RunCase(t *testing.T) {
+	testCases := []struct {
+		name       string
+		template   string
+		constraint string
+		object     string
+		assertions []Assertion
+		want       CaseResult
+	}{
+		// Validation successful
+		{
+			name:       "implicit allow",
+			template:   templateAlwaysValidate,
+			constraint: constraintAlwaysValidate,
+			object:     object,
+			assertions: nil,
+			want:       CaseResult{},
+		},
+		{
+			name:       "explicit allow boolean",
+			template:   templateAlwaysValidate,
+			constraint: constraintAlwaysValidate,
+			object:     object,
+			assertions: []Assertion{{
+				Violations: intStrFromStr("no"),
+			}},
+			want: CaseResult{},
+		},
+		{
+			name:       "implicit deny fail",
+			template:   templateAlwaysValidate,
+			constraint: constraintAlwaysValidate,
+			object:     object,
+			assertions: []Assertion{{}},
+			want: CaseResult{
+				Error: ErrUnexpectedNoViolations,
+			},
+		},
+		{
+			name:       "explicit deny boolean fail",
+			template:   templateAlwaysValidate,
+			constraint: constraintAlwaysValidate,
+			object:     object,
+			assertions: []Assertion{{
+				Violations: intStrFromStr("yes"),
+			}},
+			want: CaseResult{
+				Error: ErrUnexpectedNoViolations,
+			},
+		},
+		{
+			name:       "allow int",
+			template:   templateAlwaysValidate,
+			constraint: constraintAlwaysValidate,
+			object:     object,
+			assertions: []Assertion{{
+				Violations: intStrFromInt(0),
+			}},
+			want: CaseResult{},
+		},
+		{
+			name:       "deny int fail",
+			template:   templateAlwaysValidate,
+			constraint: constraintAlwaysValidate,
+			object:     object,
+			assertions: []Assertion{{
+				Violations: intStrFromInt(1),
+			}},
+			want: CaseResult{
+				Error: ErrUnexpectedNoViolations,
+			},
+		},
+		{
+			name:       "deny message fail",
+			template:   templateAlwaysValidate,
+			constraint: constraintAlwaysValidate,
+			object:     object,
+			assertions: []Assertion{{
+				Message: pointer.StringPtr("first message"),
+			}},
+			want: CaseResult{
+				Error: ErrUnexpectedNoViolations,
+			},
+		},
+		// Single violation
+		{
```

Comment:
Should we run the "implicit allow" test here, expecting success?

Reply:
Renamed tests for clarity.